contributors/design-proposals/local-pv-provisioner.md

```diff
@@ -0,0 +1,204 @@
+
+### Master
+A single goroutine processes events from all informers.
+
+On ProvisionerStatus Informer update, perform the following actions:
+1. For deletion: If a volume is in the Completed state, delete the corresponding PV from the API server and remove the entry from `DeleteOperations`.
+1. For discovery: compare `LocalVolumes` with existing PVs in the API server. If a volume exists in `LocalVolumes` but has not been persisted, and if the disk is not about to be deleted (i.e. in Requested or InProgress state), then create the PV.
+
+On PV Informer update: if the updated PV is in VolumeReleased state, add the PV to `DeleteOperations` with the Requested state.
+
+On Node deletion: delete the associated ProvisionerStatus object.
+
+**TODO**: Consider marking cache dirty when its corresponding API object is modified within master and ignoring Informer events if cache is dirty. This ensures master is always operating on the most up-to-date data. 
+
+### Considerations
+* This design has the following advantages:
+  * Because the API server persists state reliably, we can simplify our reasoning with failure modes.
+  * Retries and backoffs are no longer necessary.
+  * Provisioner state is more visible to admins.
+
+* If a node is compromised, it has access to modify CRDs representing other nodes.
+  * What if ProvisionerStatus is non-namespaced? Only admins should have access. If a compromised node has access to all non-namespaced resources, there is a bigger security issue.
+
+## Design 4: Hybrid RPC-CRD Approach
+Keep the same CRD object described in Design 3, but only master has write access to them. Workers must persist its delete state locally and report its status to master through gRPC. Master makes requests to workers by writing to CRD, and fetches worker status by polling (similar to Design 1).
+
+**TODO**: Use network policy to limit pod communication to only within the provisioner system.
+
+### Considerations
+* It's not safe against man-in-the-middle attacks as is. Requires encryption.
+
+## Design 5: Worker as Jobs, Watcher in DaemonSet
+
+### Overview
+One of the major issues with previous approaches is worker identity - when a worker tries to communicate with master, it has to prove it's the worker it claims to be. To fix these issues, each worker must maintain different credentials.
+
+Alternatively, master can actively create worker Jobs to perform discovery and delete rather than having long-running worker pods in a DaemonSet sending data to the master. To trigger discovery, we also have local volume directory watcher pods running in a DaemonSet.
+
+### Worker
+Workers are spawned as Jobs by master on demand. In discovery mode, a worker writes the list of local volumes in JSON format to a file inside an emptyDir. A sidecar container then reads the file and writes to its STDOUT. In deletion mode, the worker executes async delete and uses the status of the Job as an indication of deletion state.
```

Comment:
It looks like the implementation will be supporting gRPC. Is it not possible for the worker to report back the JSON via gRPC? The job could be launched with a secret provided to it (via environment variable or file mounting) and that secret could be used to prove the authenticity of the gRPC. The secrets could be "one time use", with the secrets being created and destroyed along with the creation and destruction of jobs.

Just something to consider if stdout parsing turns out to be flakey.

Reply:
According to the documentation, secrets can be read by any pod in the same namespace.

Reply:
Summarizing a bit, this design is basically providing semantics of an API object that can only be updated by a single node.  We are basically implementing a specialized API server, doing our own marshaling and data transport.

How painful would this alternative be?  Turn the master into an operator that creates a namespace + CRD per node.  Would it be annoying for users to see one namespace per node when they list namespaces?  The namespaces could be labeled so at least they can be filtered out.  This alternative means implementing a specialized DaemonSet controller, but you get to communicate through the normal API methods.

Reply:
My mistake for not including this option in the doc. My concern is in the presence of master and network failures, the job has to stay alive and keep on retrying, until it receives an ACK from master. This adds some additional complexity, and we also lose the benefit of relying on Job status as task completion status because the job won't be Complete until master acks. In comparison, I think the stdout parsing approach is simpler.

Reply:
You also mentioned the grpc alternative for parsing discovery data further down in the doc.  Another complication is you have to add authentication if you do that.

Reply:
@msau42 wouldn't each worker still need different credentials even if each CRD is in its own namespace?

Reply:
Each worker would need its own namespace, and the rolebindings to modify the CRD in their namespace.

Reply:
Right, and a different service account for each rolebinding. Credentials injection would be painful

Reply:
Namespace per node isn’t going to scale very well on large dense clusters.

We have a long term direction for containers to give them unique identities that are node specific.  We will be doing that for daemonsets specifically to address “per node identity”.  The remaining problem is scoping a nodes actions to a specific subset of resources, which seems like something sig-auth should help with.

Reply:
Yes, pod identity per node + restricting a subset of resources per pod is exactly what we need. 

But it sounds like we won't have solutions in the near future.  

So I guess the question is should we wait until those solutions exist, or proceed with the current proposal and potentially change again later

Reply:
Given that we'd like to have an implementation this quarter, I would proceed with the current proposal and change it later. I'll keep in mind while implementing that the design could potentially change.

Reply:
I would like to avoid drastically changing the design now, and then to have to drastically change it again later.  So a design that can align well with future plans would be ideal, but it also depends on how far away "future" is.

If the future plan is to be able to have per node identity + restrict resource scope for DaemonSets, can we actually just keep our current DaemonSet implementation?  We would need to understand the details of these future features to evaluate if it can address our use case.

Assuming that the API scope issues are solved, do we still gain other benefits by moving to a master/worker model?

Reply:
Just trying see if we see the same gains here:
1)  If node identity issues can be solved in the long term, then one could actually stay with the current daemonset approach. The only functionality that needs to move is the API server calls to some central proxy. The central proxy can now trust the integrity of the commands/messages originating from the daemonset pods.

2) If resource ACLs could be refined to limit access to specific nodes, then we wouldn't need central proxy at all.

Of the two, I think (1) would have the most impact on design as the daemonset implementation stays largely the same.  On the other hand (2) would be a nice to have addition which would eliminate a central proxy completely and move that functionality of API calls (small in terms of code change) back to the daemonset pods. I would not consider changes due to (2) coming into existence to be a drastic change.

Reply:
I am also assuming above that (1) is something we know we can get much sooner than (2). I should probably clarify if thats true.

Reply:
One of the reasons why we leaned towards a master-worker approach rather than a proxy approach is because, in the future when we support dynamic provisioning, we want to have a single leader deciding which node to provision to. @msau42 is this still true, now that most of the PV node selection logic is in the scheduler?

@smarterclayton @mikedanese is there a timeline for per-node identity?

Reply:
It's my AI this quarter to propose a design for scheduling and dynamic provisioning.  I'm leaning towards having the scheduler give direct input to the PVC on which node a volume should be provisioned on. 

@verult can you make a list of requirements/use-cases we would need from per-node identity + resource scope restriction features?

Reply:
@msau42 ACK

Reply:
@dhirajh regarding the proxy approach, another potential issue is it's not immune to MITM attacks as it stands. Every worker should have a unique encryption key and encrypt its messages to the proxy.